In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import scipy as sp

from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import Normalizer
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import LabelEncoder
from sklearn.datasets import load_iris
from collections import Counter

from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
from sklearn.model_selection import cross_val_score

In [ ]:
from google.colab import drive

In [ ]:
drive.mount('/content/drive')

Mounted at /content/drive


**Implement KNN regression on IRIS dataset. Apply K=5 and predict the sepal** **length of the iris flowers.**
**Predict the test datapoint = 'sepal_width' : [3.4], 'petal_length' : [1.4],** **'petal_width' : [0.3]}**

**Actual output: 4.6**

**Finding the predicted output for above test datapoint.**

In [ ]:
dataset = pd.read_csv('/content/drive/My Drive/Iris (1).csv')
print(dataset)

      Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
0      1            5.1           3.5            1.4           0.2   
1      2            4.9           3.0            1.4           0.2   
2      3            4.7           3.2            1.3           0.2   
3      4            4.6           3.1            1.5           0.2   
4      5            5.0           3.6            1.4           0.2   
..   ...            ...           ...            ...           ...   
145  146            6.7           3.0            5.2           2.3   
146  147            6.3           2.5            5.0           1.9   
147  148            6.5           3.0            5.2           2.0   
148  149            6.2           3.4            5.4           2.3   
149  150            5.9           3.0            5.1           1.8   

            Species  
0       Iris-setosa  
1       Iris-setosa  
2       Iris-setosa  
3       Iris-setosa  
4       Iris-setosa  
..              ...  
145  

In [ ]:
dataset.shape

(150, 6)

In [ ]:
dataset.head(5)

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

In [ ]:
dataset.describe()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
count  150.000000     150.000000    150.000000     150.000000    150.000000
mean    75.500000       5.843333      3.054000       3.758667      1.198667
std     43.445368       0.828066      0.433594       1.764420      0.763161
min      1.000000       4.300000      2.000000       1.000000      0.100000
25%     38.250000       5.100000      2.800000       1.600000      0.300000
50%     75.500000       5.800000      3.000000       4.350000      1.300000
75%    112.750000       6.400000      3.300000       5.100000      1.800000
max    150.000000       7.900000      4.400000       6.900000      2.500000

In [ ]:
dataset.groupby('Species').size()

Species
Iris-setosa        50
Iris-versicolor    50
Iris-virginica     50
dtype: int64

In [ ]:
feature_columns = ['SepalWidthCm','PetalLengthCm','PetalWidthCm']
X = dataset[feature_columns].values
y = dataset['SepalLengthCm'].values

In [ ]:
le = LabelEncoder()
y = le.fit_transform(y)

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [ ]:
knnr = KNeighborsRegressor(n_neighbors=5)
knnr.fit(X_train, y_train)
y_pre = knnr.predict(X_test)

In [ ]:
print("Training Accuracy = ", np.around(knnr.score(X_train, y_train)*100,3))
print("Testing Accuracy = ", np.around(knnr.score(X_test, y_test)*100,3))

Training Accuracy =  88.796
Testing Accuracy =  74.286


In [ ]:
new_datapoint = { 'sepal_width' : [3.4],'petal_length' : [1.4], 'petal_width' : [0.3]}
test_datapoint=pd.DataFrame(data=new_datapoint)
predict_testdatapoint = knnr.predict(test_datapoint)
print('The Predicted Sepal Length (cm) is:', predict_testdatapoint[0])

The Predicted Sepal Length (cm) is: 7.0


/usr/local/lib/python3.7/dist-packages/sklearn/base.py:444: UserWarning: X has feature names, but KNeighborsRegressor was fitted without feature names
  f"X has feature names, but {self.__class__.__name__} was fitted without"


We have obtained the training and testing accuracy from the KNN regression model. There is a considerable difference between both the accuracies. So we can say that the model is not fitted correctly. The predicted output of the sepal length(cm) is 7.0 cm. But the actual output should be 4.6 cm. That means the model is not correctly predicting the output of the given test data point.

In [ ]:
print('This code is contributed by Guntaka Geethika Lakshmi')
print('Registration Number - 20BCI7160')

This code is contributed by Guntaka Geethika Lakshmi
Registration Number - 20BCI7160


**Apply the kNN classifier algorithm. Apply K=5 and predict the flower of the** **given datapoint.**

**Predict the test datapoint ={'sepal length' : [4.6], 'sepal_width' : [3.4],** **'petal_length' : [1.4], 'petal_width' : [0.3], }**

**Actual output: 'flower' : ‘setosa’ Finding the predicted output for above test**  **datapoint**

In [ ]:
# Load Iris Data
iris = load_iris()

In [ ]:
iris_dataframe = pd.DataFrame(data= iris.data, columns= iris.feature_names)
print(iris_dataframe)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


In [ ]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [ ]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [ ]:
target_df = pd.DataFrame(data= iris.target, columns= ['species'])

In [ ]:
target_df = pd.DataFrame(data= iris.target, columns= ['species'])
def converter(specie):
    if specie == 0:
        return 'setosa'
    elif specie == 1:
        return 'versicolor'
    else:
        return 'virginica'
target_df['species'] = target_df['species'].apply(converter)
target_df

species
0       setosa
1       setosa
2       setosa
3       setosa
4       setosa
..         ...
145  virginica
146  virginica
147  virginica
148  virginica
149  virginica

[150 rows x 1 columns]

In [ ]:
iris_dataframe = pd.concat([iris_dataframe, target_df], axis= 1)
iris_dataframe

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

       species  
0       setosa  
1       setosa  
2       setosa  
3       setosa  
4       setosa  
..         ...  
145  virginica  
146  virginica  
147  virginica  
148  virginica  
149  virginica  

[150 rows x 5 columns]

In [ ]:
iris_dataframe.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)  
count        150.000000  
mean           1.199333  
std            0.762238  
min            0.100000  
25%            0.300000  
50%            1.300000  
75%            1.800000  
max            2.500000

In [ ]:
target_df = pd.DataFrame(columns= ['species'], data= iris.target)
target_df

species
0          0
1          0
2          0
3          0
4          0
..       ...
145        2
146        2
147        2
148        2
149        2

[150 rows x 1 columns]

In [ ]:
X = iris_dataframe[iris.feature_names].values
y = iris_dataframe['species'].values

In [ ]:
le = LabelEncoder()
y = le.fit_transform(y)

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [ ]:
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

In [ ]:
accuracy = accuracy_score(y_test, y_pred)*100
print('Accuracy of our model is ' + str(round(accuracy, 2)) + ' %.')

Accuracy of our model is 96.67 %.


In [ ]:
cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(cm)
cr = classification_report(y_test, y_pred)
print("Classification Report:",)
print (cr)

Confusion Matrix:
[[11  0  0]
 [ 0 12  1]
 [ 0  0  6]]
Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        11
           1       1.00      0.92      0.96        13
           2       0.86      1.00      0.92         6

    accuracy                           0.97        30
   macro avg       0.95      0.97      0.96        30
weighted avg       0.97      0.97      0.97        30



In [ ]:
new_datapoint = {'sepal_length' : [4.6], 'sepal_width' : [3.4], 'petal_length' : [1.4], 'petal_width' : [0.3] }
test_datapoint=pd.DataFrame(data=new_datapoint)
predict_testdatapoint = classifier.predict(test_datapoint)
predict = predict_testdatapoint[0]
print('The predicted flower for the given data point from iris dataset is',iris.target_names[predict])

The predicted flower for the given data point from iris dataset is setosa


/usr/local/lib/python3.7/dist-packages/sklearn/base.py:444: UserWarning: X has feature names, but KNeighborsClassifier was fitted without feature names
  f"X has feature names, but {self.__class__.__name__} was fitted without"


We have obtained the accuracy from the KNN classifier model. The obtained accuracy is 96.67 %. It is good accuracy. So we can say that the model is fitted correctly. The predicted output is the setosa and the actual output is also setosa. That means the model is correctly predicting the output of the given test data point.